[FIX] account: Call onchange methods if tax_line has been created previously.

victoralmau opened the pull request:
**Description of the issue/feature this PR addresses**:
From this commit https://github.com/odoo/odoo/commit/8d9788247c1b2373110caab878300426a96f4737 when creating an invoice from a purchase order with defined taxes the invoice is created without taxes.

**Impacted versions**:
- 13.0
- 14.0
- 15.0

cc @Tecnativa TT32920


--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/79720).

MiquelRForgeFlow reviewed (approved):
Error detected some time ago but didn't found how to fix it. Thanks!

wtaferner reviewed (changes requested):
Actually, it should be the other way round...

Line 612 is setting tax_line, but a bug fix just skipped this assignment and inconsistently used the wrong variable...

pedrobaeza commented:
@smetl @guva-odoo 8d97882 broke this behavior

Can you please check if this is the good patch and merge it to solve it?

victoralmau commented:
> Actually, it should be the other way round...
> 
> Line 612 is setting `tax_line`, but a bug fix [8d97882](https://github.com/odoo/odoo/commit/8d9788247c1b2373110caab878300426a96f4737) just skipped this assignment from former implementation and inconsistently used the wrong variable...

You're right. I have made some changes to use `tax_line` everywhere and make the code more consistent.

smetl commented:
@robodoo r+

william-andre reviewed (commented):
> Actually, it should be the other way round...
> 
> Line 612 is setting `tax_line`, but a bug fix [8d97882](https://github.com/odoo/odoo/commit/8d9788247c1b2373110caab878300426a96f4737) just skipped this assignment from former implementation and inconsistently used the wrong variable...

I'm not sure I agree with the other way around thing. It would feel cleaner if `taxes_map_entry` was updated somehow. Doesn't matter that much since it doesn't *seem* to be used though.

______

And the commit message could be clearer, we can't see from the diff what is actually being changed and why it didn't work.

smetl commented:
@victoralmau To be consistent with what is done on master, I prefer you first version:
https://github.com/odoo/odoo/commit/1ad895f2cfd348951247098b8dc976fb4235ef78

Fyi, the master code looks like this:
https://github.com/odoo/odoo/blob/master/addons/account/models/account_move.py#L812

The `tax_line` variable can be completely removed. Also, checking `taxes_map_entry['tax_line']` on the last condition seems useless

wtaferner commented:
@william-andre @smetl 
I agree that it is valid and probably more safe to do it the other way although it must be consistent which was my only point. If it is more readable is another story, but thx for taking it seriously to get it right :-) 

I wished that the fix which was introducing the bug would have been reviewed better as this is/was a huge issue for a stable version.

victoralmau commented:
> @victoralmau To be consistent with what is done on master, I prefer you first version: [1ad895f](https://github.com/odoo/odoo/commit/1ad895f2cfd348951247098b8dc976fb4235ef78)
> 
> Fyi, the master code looks like this: https://github.com/odoo/odoo/blob/master/addons/account/models/account_move.py#L812
> 
> The `tax_line` variable can be completely removed. Also, checking `taxes_map_entry['tax_line']` on the last condition seems useless

I have changed the code to go back to the initial change.
I have removed the `tax_line` condition.
Although the change https://github.com/odoo/odoo/blob/master/addons/account/models/account_move.py#L812 seems more optimal, in v13 it has side effects and causes the total taxes to be incorrect.

wtaferner commented:
@victoralmau 
I think they meant to replace all the `tax_line` variables in the loop accordingly (consistent use of the value of tax line instead of going back to your initial fix.

smetl commented:
@robodoo r+
